### PR TITLE
fix(hooks): block the GIT_CONFIG_KEY_<n> hooksPath bypass and stop colliding with `bind`

### LINKED ISSUE #2466: Lisa's block-no-verify.sh misses the GIT_CONFIG_KEY_<n>=core.hooksPath bypass that propswap's guard blocks
Found during WU-L fleet migration (see `plans/improvement-notes-implementation.md`).

**Applying Lisa 3.0.0 to `<consumer-repo>` would have DOWNGRADED a security guard.**

That repo hardened its own `scripts/lisa-hooks/block-no-verify.sh` to block the `GIT_CONFIG_KEY_<n>=core.hooksPath` bypass — an env-var-style git config override that defeats hook enforcement (their commit `797aa423`, with a test spawning the hook directly). **Upstream Lisa does not block it.**

Since #2436 made Lisa-owned artifacts refresh automatically on apply, the upstream (weaker) guard would have silently overwritten the local (stronger) one. Their own tests caught it and the migration excluded the incidental refresh — but this is exactly the failure mode where an improvement to delivery makes a security regression easier to ship.

**Two actions:**
1. Port the `GIT_CONFIG_KEY_<n>=core.hooksPath` protection upstream into `block-no-verify.sh` (also check the case-variation and unpinned-`npx` hardening from `9ae9154`).
2. Consider whether the Lisa-owned refresh path should detect that a host's copy is *stronger* rather than merely different — a pure byte comparison cannot tell a downgrade from an update.

Parent: CodySwannGT/lisa#2423

### PR DESCRIPTION
**Pull request by CodySwannGT**
Fixes two guard defects of the same shape: a check that matched a **spelling** rather than the thing it meant to catch.

Work-Item: CodySwannGT/lisa#2466

Closes #2466, closes #2463 (`Closes` does not fire in this repo — closing by hand after merge).

## 1. `GIT_CONFIG_KEY_<n>=core.hooksPath` bypass (security)

`block-no-verify.sh` caught `-c core.hooksPath=...` and `--config-env=core.hooksPath=...`, but not the env-var-style equivalent:

```
GIT_CONFIG_COUNT=1 GIT_CONFIG_KEY_0=core.hooksPath GIT_CONFIG_VALUE_0=/dev/null git commit
```

That sets the same command-scope config and disables **every** git hook just as completely, while matching none of the token shapes the guard knew about.

`propswapllc/frontend` had already hardened its own copy (`797aa423`). Upstream had not — so after #2436 made Lisa-owned artifacts refresh on apply, the next fleet update would have **silently overwritten the stronger downstream guard with the weaker upstream one**. Their tests caught it during the WU-L migration; nothing else would have.

Design notes:
- The index is matched as `\d+`, **not pinned to 0** — a single-index check is evaded by typing a `1`.
- The key is refused **outright**, like `--config-env=`, because the path lives in a separate `GIT_CONFIG_VALUE_<n>` token that can be exported earlier, reordered, or omitted. There is nothing to allowlist against.
- Case-insensitive, since git config names are.

### Cross-harness parity gap (found while here)

The Codex and Antigravity variants had fallen behind the Claude one. Both still:
- denylisted only `""` and `/dev/null`, so `-c core.hooksPath=/tmp/empty` passed;
- matched config names case-sensitively, so `CORE.HOOKSPATH=` passed;
- knew nothing of `--config-env=`.

A bypass that survives by switching harness is still a bypass, so all three are now at parity, each with its **own** test suite rather than one suite standing in for the rest.

## 2. `lisa-work-item.mjs bind` unrunnable in a worktree

`bind` names a shell builtin that evaluates a string, so Claude Code's worktree isolation refuses any command line containing the bare token — **wherever** it appears, including as a subcommand three words in. That made the one step Lisa documents as mandatory before durable work impossible to run in the isolated worktrees Lisa itself tells agents to use. Three agents hit it in one session, each rediscovering the same hand-written-trailer workaround.

**The guard is in the Claude Code binary** (v2.1.229, `strings` confirms), not in this repo — it is not Lisa's to narrow, and it is doing something legitimate. So Lisa stops standing on its toes instead:

- `link` is now the documented spelling and does identical work.
- `bind` stays a **permanent alias**, not a deprecation: host projects' checked-in hooks and CI already say it, and removing it would break the very traceability gate the file exists to enforce, in repos that did nothing wrong.

Docs, rules, and skills across all six agent surfaces now teach `link`, with a note explaining why.

## Test evidence

- `tests/unit/hooks/block-no-verify.test.ts` — 13 new RED cases (numbered indices 0/1/2/7/10/42, case variants, quoted, via `env`, allowlisted-value-still-denied), all failing before the fix.
- `tests/unit/codex/block-no-verify-codex.test.ts` — new suite, spawns the Codex hook directly.
- `tests/unit/agy/block-no-verify-agy.test.ts` — parity cases added.
- `tests/unit/scripts/lisa-work-item.test.ts` — `link` works, `bind` still works, usage names `link`.
- Full suite: **10,741 passing**. `lint` and `typecheck` clean.
- End-to-end: this PR's own work item was bound with `node scripts/lisa-work-item.mjs link CodySwannGT/lisa#2466` **from inside an isolated worktree** — the exact command that was blocked.

## Judgement requested: downgrade detection

Lisa-owned refresh is a **byte comparison**, which cannot tell a host's *stronger* guard from stale drift — that is the root cause that made defect 1 dangerous rather than merely missing. Fixing the guard does not fix that.

I did **not** build it here (it is a separate subsystem touching apply/refresh, `lisa doctor`, and every Lisa-owned artifact, and it needs its own design review). Recommended shape, cheapest first:

1. **Capability marker in the guard header** — a `# lisa-guard-capabilities: no-verify-abbrev, hookspath-allowlist, config-env, git-config-key` line. Refresh parses both sides and refuses to overwrite when the host declares a **superset**. Cheap, self-documenting, reviewable in a diff, and degrades safely (no marker = today's behavior).
2. **Monotonic guard version** — simpler to compare but lies the moment two hosts harden different vectors independently, which is exactly what happened here (propswap and tunnl hardened different things).

Option 1 is the one I would file. Happy to open it as a ticket.

🤖 Generated with Claude Code

## Summary by CodeRabbit

* **New Features**
  * Added `link` as the primary work-item association command; `bind` remains supported as an alias.
  * Improved validation of Git hook paths, allowing approved project hook directories while rejecting unsafe configurations.
  * Added protection against additional environment-based methods of bypassing Git hooks.

* **Bug Fixes**
  * Closed multiple bypass paths that could disable hook checks through Git configuration.

* **Tests**
  * Expanded coverage for hook protection, work-item linking, aliases, and configuration edge cases.

### CODE REVIEW

File: plugins/src/base/hooks/block-no-verify.agy.sh [hunk 1]
```diff
@@ -102,16 +102,56 @@ def disables_verification(token):
...
+            sys.exit(1)
+    if lowered == "core.hookspath" and i + 1 < len(normalized_tokens):
+        if not is_permitted_hooks_path(normalized_tokens[i + 1]):
+            sys.exit(1)
+    # `--config-env=core.hooksPath=SOMEVAR` reads the path out of the named env
+    # var, so it is not in the command at all and cannot be allowlisted.
+    if lowered.startswith("--config-env="):
+        spec = token.split("=", 1)[1]
+        if spec.split("=", 1)[0].strip().lower() == "core.hookspath":
             sys.exit(1)
```
[the remark this answers is not in the document]
Comment (CodySwannGT) [reply plugins/src/base/hooks/block-no-verify.agy.sh:148]: Valid, and a genuinely good catch — fixed in c7d7b242a.

Reproduced against real git before fixing:

```
$ env '.husky=/dev/null' git -C "$repo" --config-env core.hooksPath=.husky config --get core.hooksPath
/dev/null
```

The failure was worse than "one spelling unguarded". Because only the `=` form was matched, the separate form's trailing `core.hooksPath=.husky` fell through to the hooksPath **allowlist**, which reads `.husky` as a *path* and permits it. In that position it is an environment **variable name**, which can hold anything — so the allowlist added to make this guard stronger was itself the bypass. Confirmed the pre-fix hook exited 0 on `git --config-env core.hooksPath=.husky commit -m bypass`.

Fixed at the `--config-env` token rather than the value token, since the loop reaches it first and can refuse before the next token can be mistaken for a path. Applied to the three **source** variants — `plugins/src/base/hooks/block-no-verify.sh`, `plugins/src/base/hooks/block-no-verify.agy.sh`, `src/codex/scripts/block-no-verify.sh`. The other three files you listed (`plugins/lisa/`, `plugins/lisa-cursor/`, `plugins/lisa-copilot/`, and `all/copy-overwrite/`) are build outputs of `scripts/build-plugins.sh` and a unit test asserts byte-equality with the source, so editing them directly would have been reverted by the next build; they picked the fix up via `bun run build:plugins`.

Regression coverage added to all three suites, parameterized over `.husky`, `.githooks`, and `HOOKS` — the two permitted names are exactly the ones that got through, so they are the ones worth pinning.

File: tests/unit/scripts/lisa-work-item.test.ts [hunk 2]
```diff
@@ -270,6 +270,11 @@ function wedgeRebase(fixture: Fixture, branch: string): void {
     throw new Error("unexpected rebase head-name");
 }
 
+/** Path to the worktree-private binding state a successful bind/link writes. */
+function stateFilePath(fixture: Fixture): string {
+  return path.join(fixture.root, ".git", "lisa", "work-item.json");
+}
```
[the remark this answers is not in the document]
Comment (CodySwannGT) [reply tests/unit/scripts/lisa-work-item.test.ts]: Valid — fixed in c7d7b242a.

`stateFilePath()` now resolves through `git rev-parse --git-path lisa/work-item.json`, the same call the CLI itself makes, and the hardcoded assertion at the old line 412 uses the helper instead of rebuilding the path by hand.

Worth noting why this mattered beyond tidiness: in a **linked** worktree the state file is not at `.git/lisa/` at all — it lands in `.git/worktrees/<name>/lisa/`. I hit that directly while testing this PR; binding this PR's own work item wrote to `.git/worktrees/agent-a47af7cc2c2395873/lisa/work-item.json`. The fixtures are plain repos so the hardcoded path passed, but it would have silently stopped describing the thing under test in exactly the environment this binding exists to serve.
